executor/compact_table.go

```diff
@@ -0,0 +1,337 @@
+			Type:    tikvrpc.CmdCompact,
+			StoreTp: tikvrpc.TiFlash,
+			Req: &kvrpcpb.CompactRequest{
+				LogicalTableId:  task.parentExec.tableInfo.ID,
+				PhysicalTableId: physicalTableID,
+				StartKey:        startKey,
+			},
+		})
+		if err != nil {
+			// Even after backoff, the request is still failed.., or the request is cancelled or timed out
+			warn := errors.Errorf("compact on store %s failed: %v", task.targetStore.Address, err)
+			task.parentExec.ctx.GetSessionVars().StmtCtx.AppendWarning(warn)
+			task.logFailure(
+				zap.Int64("physical-table-id", physicalTableID),
+				zap.Error(err))
+			return false, warn
+		}
+		if resp.GetError() != nil {
+			switch resp.GetError().GetError().(type) {
+			case *kvrpcpb.CompactError_ErrCompactInProgress:
+				warn := errors.Errorf("compact on store %s failed: table is compacting in progress", task.targetStore.Address)
+				task.parentExec.ctx.GetSessionVars().StmtCtx.AppendWarning(warn)
+				task.logFailure(
+					zap.Int64("physical-table-id", physicalTableID),
+					zap.Error(warn))
+				// TiFlash reported that there are existing compacting for the same table.
+				// We should stop the whole SQL execution, including compacting requests to other stores, as repeatedly
+				// compacting the same table is a waste of resource.
+				return true, warn
+			case *kvrpcpb.CompactError_ErrTooManyPendingTasks:
+				// The store is already very busy, don't retry.
+				warn := errors.Errorf("compact on store %s failed: store is too busy", task.targetStore.Address)
+				task.parentExec.ctx.GetSessionVars().StmtCtx.AppendWarning(warn)
+				task.logFailure(
+					zap.Int64("physical-table-id", physicalTableID),
+					zap.Error(warn))
+				return false, warn
+			default:
+				// Others are unexpected errors, don't retry.
+				warn := errors.Errorf("compact on store %s failed: internal error (check logs for details)", task.targetStore.Address)
+				task.parentExec.ctx.GetSessionVars().StmtCtx.AppendWarning(warn)
+				task.logFailure(
+					zap.Int64("physical-table-id", physicalTableID),
+					zap.Any("response-error", resp.GetError().GetError()))
+				return false, warn
+			}
+		}
+
+		if !resp.HasRemaining {
+			return false, nil
+		}
+
+		// Let's send more compact requests, as there are remaining data to compact.
+		lastEndKey := resp.GetCompactedEndKey()
+		if len(lastEndKey) == 0 || bytes.Compare(lastEndKey, startKey) <= 0 {
+			// The TiFlash server returned an invalid compacted end key.
+			// This is unexpected...
+			warn := errors.Errorf("compact on store %s failed: internal error (check logs for details)", task.targetStore.Address)
+			task.parentExec.ctx.GetSessionVars().StmtCtx.AppendWarning(warn)
+			task.logFailure(
+				zap.Int64("physical-table-id", physicalTableID),
+				zap.String("compacted-start-key", hex.EncodeToString(resp.GetCompactedStartKey())),
+				zap.String("compacted-end-key", hex.EncodeToString(resp.GetCompactedEndKey())),
+			)
+			return false, warn
+		}
```

Comment:
is this possible?

Reply:
According to the TiFlash logic it is not possible. Actually there is also an assertion at TiFlash side to ensure this will not happen.

This is just a protection, so that when it indeed happens we will instantly know where it happens, instead of letting the problem propagating to elsewhere.